core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala

```diff
@@ -0,0 +1,240 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.internal.plugin
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.util.{Map => JMap}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+import com.codahale.metrics.Gauge
+import com.google.common.io.Files
+import org.mockito.ArgumentMatchers.{any, eq => meq}
+import org.mockito.Mockito.{mock, spy, verify, when}
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.concurrent.Eventually.{eventually, interval, timeout}
+
+import org.apache.spark.{ExecutorPlugin => _, _}
+import org.apache.spark.api.plugin._
+import org.apache.spark.internal.config._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.util.Utils
+
+class PluginContainerSuite extends SparkFunSuite with BeforeAndAfterEach with LocalSparkContext {
+
+  override def afterEach(): Unit = {
+    TestSparkPlugin.reset()
+    super.afterEach()
+  }
+
+  test("plugin initialization and communication") {
+    val conf = new SparkConf()
+      .setAppName(getClass().getName())
+      .set(SparkLauncher.SPARK_MASTER, "local[1]")
+      .set(PLUGINS, Seq(classOf[TestSparkPlugin].getName()))
+
+    TestSparkPlugin.extraConf = Map("foo" -> "bar", "bar" -> "baz").asJava
+
+    sc = new SparkContext(conf)
+
+    assert(TestSparkPlugin.driverPlugin != null)
+    verify(TestSparkPlugin.driverPlugin).init(meq(sc), any())
+
+    assert(TestSparkPlugin.executorPlugin != null)
+    verify(TestSparkPlugin.executorPlugin).init(any(), meq(TestSparkPlugin.extraConf))
+
+    assert(TestSparkPlugin.executorContext != null)
+
+    // One way messages don't block, so need to loop checking whether it arrives.
+    TestSparkPlugin.executorContext.send("oneway")
+    eventually(timeout(10.seconds), interval(10.millis)) {
+      verify(TestSparkPlugin.driverPlugin).receive("oneway")
+    }
+
+    assert(TestSparkPlugin.executorContext.ask("ask") === "reply")
+
+    val err = intercept[Exception] {
+      TestSparkPlugin.executorContext.ask("unknown message")
+    }
+    assert(err.getMessage().contains("unknown message"))
+
+    // It should be possible for the driver plugin to send a message to itself, even if that doesn't
+    // make a whole lot of sense. It at least allows the same context class to be used on both
```

Comment:
this answers my question about this.  I wasn't sure it made much sense but seems ok for reuse. Was wondering if you had a specific use case in mind but sounds like not.